src/lib/logical_query_plan/join_node.cpp

```diff
@@ -228,19 +230,72 @@ bool JoinNode::is_column_nullable(const ColumnID column_id) const {
 
 const std::vector<std::shared_ptr<AbstractExpression>>& JoinNode::join_predicates() const { return node_expressions; }
 
+void JoinNode::mark_as_semi_reduction(const std::shared_ptr<JoinNode>& reduced_join_node) {
+  Assert(!_is_semi_reduction, "The semi reducer status should be set once only.");
+  Assert(reduced_join_node, "Reduced JoinNode must be provided.");
+  Assert(join_mode == JoinMode::Semi, "Semi join reductions require JoinMode::Semi.");
+  DebugAssert(join_predicates().size() == 1,
+              "Currently, semi join reductions are expected to have a single join predicate.");
+  DebugAssert(std::any_of(reduced_join_node->join_predicates().begin(), reduced_join_node->join_predicates().end(),
+                          [&](const auto predicate) { return *predicate == *join_predicates()[0]; }),
+              "Both semi join reduction node and the reduced join should have a common join predicate.");
+  _is_semi_reduction = true;
+  _reduced_join_node = std::weak_ptr<JoinNode>(reduced_join_node);
+}
+
+bool JoinNode::is_semi_reduction() const {
+  DebugAssert(!_is_semi_reduction || join_mode == JoinMode::Semi, "Non-semi join is marked as a semi reduction.");
+  return _is_semi_reduction;
+}
+
+std::shared_ptr<JoinNode> JoinNode::get_or_find_reduced_join_node() const {
+  if (!_is_semi_reduction) {
```

Comment:
What is the semantic of returning a `nullptr` when the node is not a reducer? I'd expect an error (aka `Assert`) in that case

Reply:
Yes, you are right!